File: src/data_dirs.rs
//! Provider data-directory discovery (mirrors Go `internal/datadirs`).
//!
//! Every supported provider is one row in [`PROVIDERS`]; the per-provider
//! quirks (comma-separated env overrides, Claude's trailing `projects`
//! segment, Goose's session-db suffix) are data on the row, not code paths.
//! There is no provider selection UI — all providers, always.

use std::{
    collections::{BTreeMap, BTreeSet},
    fs,
    path::{Path, PathBuf},
};

/// Resolved provider directories: provider name → existing data directories.
pub type ProviderDirs = BTreeMap<String, Vec<PathBuf>>;

struct Provider {
    name: &'static str,
    env: &'static str,
    /// Appended to every env-provided root (Goose's `sessions.db` path).
    env_suffix: &'static str,
    /// Trim a trailing `projects` segment from env entries (Claude quirk).
    trim_projects: bool,
    /// Default candidates; `~/` is the home dir, `$XDG_CONFIG_HOME/` expands
    /// with a `~/.config` fallback.
    defaults: &'static [&'static str],
}

const PROVIDERS: &[Provider] = &[
    Provider {
        name: "claude",
        env: "CLAUDE_CONFIG_DIR",
        env_suffix: "",
        trim_projects: true,
        defaults: &["~/.claude", "$XDG_CONFIG_HOME/claude"],
    },
    Provider {
        name: "codex",
        env: "CODEX_CONFIG_DIR",
        env_suffix: "",
        trim_projects: false,
        defaults: &["~/.codex"],
    },
    Provider {
        name: "copilot",
        env: "COPILOT_OTEL_FILE_EXPORTER_PATH",
        env_suffix: "",
        trim_projects: false,
        defaults: &["~/.copilot/otel"],
    },
    Provider {
        name: "gemini",
        env: "GEMINI_DATA_DIR",
        env_suffix: "",
        trim_projects: false,
        defaults: &["~/.gemini/tmp"],
    },
    Provider {
        name: "kimi",
        env: "KIMI_DATA_DIR",
        env_suffix: "",
        trim_projects: false,
        defaults: &["~/.kimi"],
    },
    Provider {
        name: "qwen",
        env: "QWEN_DATA_DIR",
        env_suffix: "",
        trim_projects: false,
        defaults: &["~/.qwen"],
    },
    Provider {
        name: "openclaw",
        env: "OPENCLAW_DIR",
        env_suffix: "",
        trim_projects: false,
        defaults: &["~/.openclaw", "~/.clawdbot", "~/.moltbot", "~/.moldbot"],
    },
    Provider {
        name: "pi",
        env: "PI_AGENT_DIR",
        env_suffix: "",
        trim_projects: false,
        defaults: &["~/.pi/agent/sessions"],
    },
    Provider {
        name: "amp",
        env: "AMP_DATA_DIR",
        env_suffix: "",
        trim_projects: false,
        defaults: &["~/.local/share/amp"],
    },
    Provider {
        name: "droid",
        env: "DROID_SESSIONS_DIR",
        env_suffix: "",
        trim_projects: false,
        defaults: &["~/.factory/sessions"],
    },
    Provider {
        name: "kilo",
        env: "KILO_DATA_DIR",
        env_suffix: "",
        trim_projects: false,
        defaults: &["~/.local/share/kilo"],
    },
    Provider {
        name: "hermes",
        env: "HERMES_HOME",
        env_suffix: "",
        trim_projects: false,
        defaults: &["~/.hermes"],
    },
    Provider {
        name: "codebuff",
        env: "CODEBUFF_DATA_DIR",
        env_suffix: "",
        trim_projects: false,
        defaults: &[
            "~/.config/manicode",
            "~/.config/manicode-dev",
            "~/.config/manicode-staging",
        ],
    },
    Provider {
        name: "opencode",
        env: "OPENCODE_DATA_DIR",
        env_suffix: "",
        trim_projects: false,
        defaults: &["~/.local/share/opencode"],
    },
    Provider {
        name: "goose",
        env: "GOOSE_PATH_ROOT",
        env_suffix: "data/sessions/sessions.db",
        trim_projects: false,
        defaults: &[
            "~/.local/share/goose/sessions/sessions.db",
            "~/Library/Application Support/goose/sessions/sessions.db",
            "~/.local/share/Block/goose/sessions/sessions.db",
        ],
    },
];

/// Returns, for each provider, the first existing candidate path.
///
/// Providers with no existing candidate are omitted; an empty map means there
/// is nothing to sync.
#[must_use]
pub fn resolve() -> ProviderDirs {
    resolve_with(&|name| std::env::var(name).ok(), &home_dir())
}

/// Returns the union of watchable directories over all candidate paths of all
/// providers: an existing directory watches itself, an existing file watches
/// its parent. Deduplicated and sorted.
#[must_use]
pub fn watch_paths() -> Vec<PathBuf> {
    watch_paths_with(&|name| std::env::var(name).ok(), &home_dir())
}

fn home_dir() -> PathBuf {
    ["USERPROFILE", "HOME"]
        .iter()
        .find_map(std::env::var_os)
        .map_or_else(|| PathBuf::from("."), PathBuf::from)
}

fn resolve_with(env: &dyn Fn(&str) -> Option<String>, home: &Path) -> ProviderDirs {
    PROVIDERS
        .iter()
        .filter_map(|provider| {
            let first = candidates(provider, env, home)
                .into_iter()
                .find(|path| fs::metadata(path).is_ok())?;
            Some((provider.name.to_owned(), vec![first]))
        })
        .collect()
}

fn watch_paths_with(env: &dyn Fn(&str) -> Option<String>, home: &Path) -> Vec<PathBuf> {
    let unique: BTreeSet<PathBuf> = PROVIDERS
        .iter()
        .flat_map(|provider| candidates(provider, env, home))
        .filter_map(|path| existing_watch_dir(&path))
        .collect();
    unique.into_iter().collect()
}

fn candidates(
    provider: &Provider,
    env: &dyn Fn(&str) -> Option<String>,
    home: &Path,
) -> Vec<PathBuf> {
    if let Some(value) = env(provider.env) {
        return split_configured_dirs(&value, provider.trim_projects)
            .into_iter()
            .map(|root| {
                if provider.env_suffix.is_empty() {
                    root
                } else {
                    root.join(provider.env_suffix)
                }
            })
            .collect();
    }
    provider
        .defaults
        .iter()
        .map(|template| expand(template, env, home))
        .collect()
}

/// Splits a comma-separated directory list, trimming whitespace and dropping
/// empties. With `trim_projects`, a trailing `projects` segment is rewritten
/// to its parent (Claude stores sessions under `<config>/projects`).
fn split_configured_dirs(value: &str, trim_projects: bool) -> Vec<PathBuf> {
    value
        .split(',')
        .map(str::trim)
        .filter(|entry| !entry.is_empty())
        .map(|entry| {
            let path = PathBuf::from(entry);
            if trim_projects && path.file_name().is_some_and(|name| name == "projects") {
                path.parent().map_or(path.clone(), Path::to_path_buf)
            } else {
                path
            }
        })
        .collect()
}

fn expand(template: &str, env: &dyn Fn(&str) -> Option<String>, home: &Path) -> PathBuf {
    if let Some(rest) = template.strip_prefix("~/") {
        return home.join(rest);
    }
    if let Some(rest) = template.strip_prefix("$XDG_CONFIG_HOME/") {
        let config = env("XDG_CONFIG_HOME").map_or_else(|| home.join(".config"), PathBuf::from);
        return config.join(rest);
    }
    PathBuf::from(template)
}

fn existing_watch_dir(path: &Path) -> Option<PathBuf> {
    let meta = fs::metadata(path).ok()?;
    if meta.is_dir() {
        return Some(path.to_path_buf());
    }
    let parent = path.parent()?;
    fs::metadata(parent)
        .ok()?
        .is_dir()
        .then(|| parent.to_path_buf())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn no_env(_: &str) -> Option<String> {
        None
    }

    mod split_configured_dirs {
        use super::*;

        #[test]
        fn trims_whitespace_and_drops_empties() {
            let dirs = split_configured_dirs(" a , ,b,", false);
            assert_eq!(dirs, vec![PathBuf::from("a"), PathBuf::from("b")]);
        }

        #[test]
        fn rewrites_trailing_projects_segment_to_parent() {
            let dirs = split_configured_dirs("/home/u/.claude/projects", true);
            assert_eq!(dirs, vec![PathBuf::from("/home/u/.claude")]);
        }

        #[test]
        fn keeps_projects_segment_without_trim_flag() {
            let dirs = split_configured_dirs("/home/u/.claude/projects", false);
            assert_eq!(dirs, vec![PathBuf::from("/home/u/.claude/projects")]);
        }
    }

    mod resolve_with {
        use super::*;

        #[test]
        fn picks_first_existing_default_candidate() {
            let home = tempfile::tempdir().unwrap();
            fs::create_dir_all(home.path().join(".claude")).unwrap();

            let dirs = resolve_with(&no_env, home.path());

            assert_eq!(dirs.get("claude"), Some(&vec![home.path().join(".claude")]));
        }

        #[test]
        fn omits_providers_with_no_existing_candidate() {
            let home = tempfile::tempdir().unwrap();
            let dirs = resolve_with(&no_env, home.path());
            assert!(dirs.is_empty(), "unexpected: {dirs:?}");
        }

        #[test]
        fn env_override_replaces_defaults() {
            let home = tempfile::tempdir().unwrap();
            let custom = home.path().join("custom-codex");
            fs::create_dir_all(&custom).unwrap();
            fs::create_dir_all(home.path().join(".codex")).unwrap();
            let custom_str = custom.to_string_lossy().into_owned();
            let env = move |name: &str| (name == "CODEX_CONFIG_DIR").then(|| custom_str.clone());

            let dirs = resolve_with(&env, home.path());

            assert_eq!(dirs.get("codex"), Some(&vec![custom]));
        }

        #[test]
        fn goose_env_root_gets_sessions_db_suffix() {
            let home = tempfile::tempdir().unwrap();
            let root = home.path().join("goose-root");
            let db_dir = root.join("data").join("sessions");
            fs::create_dir_all(&db_dir).unwrap();
            fs::write(db_dir.join("sessions.db"), b"").unwrap();
            let root_str = root.to_string_lossy().into_owned();
            let env = move |name: &str| (name == "GOOSE_PATH_ROOT").then(|| root_str.clone());

            let dirs = resolve_with(&env, home.path());

            assert_eq!(dirs.get("goose"), Some(&vec![db_dir.join("sessions.db")]));
        }
    }

    mod watch_paths_with {
        use super::*;

        #[test]
        fn maps_existing_file_candidate_to_its_parent() {
            let home = tempfile::tempdir().unwrap();
            let db_dir = home.path().join(".local/share/goose/sessions");
            fs::create_dir_all(&db_dir).unwrap();
            fs::write(db_dir.join("sessions.db"), b"").unwrap();

            let paths = watch_paths_with(&no_env, home.path());

            assert_eq!(paths, vec![db_dir]);
        }

        #[test]
        fn dedupes_and_sorts_directories() {
            let home = tempfile::tempdir().unwrap();
            fs::create_dir_all(home.path().join(".claude")).unwrap();
            fs::create_dir_all(home.path().join(".codex")).unwrap();

            let paths = watch_paths_with(&no_env, home.path());

            assert_eq!(
                paths,
                vec![home.path().join(".claude"), home.path().join(".codex")]
            );
        }
    }
}

File: src/data_dirs.c
#include "data_dirs.h"

#include "util/wstr.h"

#include <windows.h>

#include <stdlib.h>
#include <string.h>

typedef struct Provider {
    const wchar_t *name;
    const wchar_t *env;
    /* Appended to every env-provided root (Goose's sessions.db path). */
    const wchar_t *env_suffix;
    /* Rewrite a trailing `projects` segment to its parent (Claude quirk). */
    bool trim_projects;
    /* Defaults: `~/` is home, `$XDG_CONFIG_HOME/` falls back to ~/.config. */
    const wchar_t *defaults[4];
} Provider;

static const Provider PROVIDERS[] = {
    {L"claude", L"CLAUDE_CONFIG_DIR", NULL, true,
     {L"~/.claude", L"$XDG_CONFIG_HOME/claude", NULL, NULL}},
    {L"codex", L"CODEX_CONFIG_DIR", NULL, false, {L"~/.codex", NULL, NULL, NULL}},
    {L"copilot", L"COPILOT_OTEL_FILE_EXPORTER_PATH", NULL, false,
     {L"~/.copilot/otel", NULL, NULL, NULL}},
    {L"gemini", L"GEMINI_DATA_DIR", NULL, false, {L"~/.gemini/tmp", NULL, NULL, NULL}},
    {L"kimi", L"KIMI_DATA_DIR", NULL, false, {L"~/.kimi", NULL, NULL, NULL}},
    {L"qwen", L"QWEN_DATA_DIR", NULL, false, {L"~/.qwen", NULL, NULL, NULL}},
    {L"openclaw", L"OPENCLAW_DIR", NULL, false,
     {L"~/.openclaw", L"~/.clawdbot", L"~/.moltbot", L"~/.moldbot"}},
    {L"pi", L"PI_AGENT_DIR", NULL, false, {L"~/.pi/agent/sessions", NULL, NULL, NULL}},
    {L"amp", L"AMP_DATA_DIR", NULL, false, {L"~/.local/share/amp", NULL, NULL, NULL}},
    {L"droid", L"DROID_SESSIONS_DIR", NULL, false,
     {L"~/.factory/sessions", NULL, NULL, NULL}},
    {L"kilo", L"KILO_DATA_DIR", NULL, false, {L"~/.local/share/kilo", NULL, NULL, NULL}},
    {L"hermes", L"HERMES_HOME", NULL, false, {L"~/.hermes", NULL, NULL, NULL}},
    {L"codebuff", L"CODEBUFF_DATA_DIR", NULL, false,
     {L"~/.config/manicode", L"~/.config/manicode-dev", L"~/.config/manicode-staging",
      NULL}},
    {L"opencode", L"OPENCODE_DATA_DIR", NULL, false,
     {L"~/.local/share/opencode", NULL, NULL, NULL}},
    {L"goose", L"GOOSE_PATH_ROOT", L"data/sessions/sessions.db", false,
     {L"~/.local/share/goose/sessions/sessions.db",
      L"~/Library/Application Support/goose/sessions/sessions.db",
      L"~/.local/share/Block/goose/sessions/sessions.db", NULL}},
};

#define MAX_CANDIDATES 16

typedef struct Candidates {
    wchar_t *paths[MAX_CANDIDATES]; /* malloc'd */
    size_t count;
} Candidates;

static bool os_env(const wchar_t *name, wchar_t *out, size_t cap) {
    DWORD len = GetEnvironmentVariableW(name, out, (DWORD)cap);
    return len > 0 && len < cap;
}

/* Converts forward slashes from table templates into backslashes. */
static wchar_t *normalize_separators(wchar_t *path) {
    for (wchar_t *p = path; *p; p++) {
        if (*p == L'/') {
            *p = L'\\';
        }
    }
    return path;
}

static wchar_t *join_path(const wchar_t *base, const wchar_t *rest) {
    size_t cap = wcslen(base) + wcslen(rest) + 2;
    wchar_t *joined = malloc(cap * sizeof(wchar_t));
    if (joined) {
        _snwprintf_s(joined, cap, _TRUNCATE, L"%s\\%s", base, rest);
        normalize_separators(joined);
    }
    return joined;
}

static wchar_t *expand_template(const wchar_t *tmpl, EnvLookupFn env, const wchar_t *home) {
    if (wcsncmp(tmpl, L"~/", 2) == 0) {
        return join_path(home, tmpl + 2);
    }
    if (wcsncmp(tmpl, L"$XDG_CONFIG_HOME/", 17) == 0) {
        wchar_t config[MAX_PATH];
        if (env(L"XDG_CONFIG_HOME", config, MAX_PATH)) {
            return join_path(config, tmpl + 17);
        }
        wchar_t *fallback = join_path(home, L".config");
        if (!fallback) {
            return NULL;
        }
        wchar_t *joined = join_path(fallback, tmpl + 17);
        free(fallback);
        return joined;
    }
    wchar_t *copy = wstr_dup(tmpl);
    return copy ? normalize_separators(copy) : NULL;
}

/* Rewrites a trailing `projects` segment to its parent, in place. */
static void trim_projects_segment(wchar_t *path) {
    size_t len = wcslen(path);
    while (len > 0 && (path[len - 1] == L'\\' || path[len - 1] == L'/')) {
        path[--len] = L'\0';
    }
    const wchar_t *last = path + len;
    while (last > path && last[-1] != L'\\' && last[-1] != L'/') {
        last--;
    }
    if (wcscmp(last, L"projects") == 0 && last > path) {
        path[(last - path) - 1] = L'\0';
    }
}

/* Splits a comma-separated env value into candidate paths. */
static void split_configured(const wchar_t *value, const Provider *p, Candidates *out) {
    const wchar_t *cursor = value;
    while (*cursor && out->count < MAX_CANDIDATES) {
        const wchar_t *comma = wcschr(cursor, L',');
        size_t len = comma ? (size_t)(comma - cursor) : wcslen(cursor);

        /* Trim whitespace. */
        while (len > 0 && (*cursor == L' ' || *cursor == L'\t')) {
            cursor++;
            len--;
        }
        while (len > 0 && (cursor[len - 1] == L' ' || cursor[len - 1] == L'\t')) {
            len--;
        }
        if (len > 0) {
            wchar_t *entry = malloc((len + 1) * sizeof(wchar_t));
            if (entry) {
                wmemcpy(entry, cursor, len);
                entry[len] = L'\0';
                normalize_separators(entry);
                if (p->trim_projects) {
                    trim_projects_segment(entry);
                }
                if (p->env_suffix) {
                    wchar_t *with_suffix = join_path(entry, p->env_suffix);
                    free(entry);
                    entry = with_suffix;
                }
                if (entry) {
                    out->paths[out->count++] = entry;
                }
            }
        }
        cursor = comma ? comma + 1 : cursor + len;
        if (comma) {
            continue;
        }
        break;
    }
}

static void candidates_for(const Provider *p, EnvLookupFn env, const wchar_t *home,
                           Candidates *out) {
    out->count = 0;
    wchar_t value[4096];
    if (env(p->env, value, 4096)) {
        split_configured(value, p, out);
        return;
    }
    for (size_t i = 0; i < 4 && p->defaults[i]; i++) {
        wchar_t *path = expand_template(p->defaults[i], env, home);
        if (path && out->count < MAX_CANDIDATES) {
            out->paths[out->count++] = path;
        } else {
            free(path);
        }
    }
}

static void candidates_free(Candidates *c) {
    for (size_t i = 0; i < c->count; i++) {
        free(c->paths[i]);
    }
    c->count = 0;
}

static bool path_exists(const wchar_t *path, bool *is_dir) {
    DWORD attrs = GetFileAttributesW(path);
    if (attrs == INVALID_FILE_ATTRIBUTES) {
        return false;
    }
    *is_dir = (attrs & FILE_ATTRIBUTE_DIRECTORY) != 0;
    return true;
}

static void default_home(wchar_t *out, size_t cap) {
    if (!os_env(L"USERPROFILE", out, cap) && !os_env(L"HOME", out, cap)) {
        wcscpy_s(out, cap, L".");
    }
}

void data_dirs_resolve_with(EnvLookupFn env, const wchar_t *home, ProviderDirs *out) {
    size_t provider_count = ARRAYSIZE(PROVIDERS);
    out->items = calloc(provider_count, sizeof(ProviderDir));
    out->count = 0;
    if (!out->items) {
        return;
    }
    for (size_t i = 0; i < provider_count; i++) {
        Candidates c;
        candidates_for(&PROVIDERS[i], env, home, &c);
        for (size_t j = 0; j < c.count; j++) {
            bool is_dir;
            if (path_exists(c.paths[j], &is_dir)) {
                out->items[out->count].provider = PROVIDERS[i].name;
                out->items[out->count].dir = c.paths[j];
                c.paths[j] = NULL; /* ownership moved */
                out->count++;
                break;
            }
        }
        candidates_free(&c);
    }
    /* Providers sorted by name, like the Rust BTreeMap ordering. */
    for (size_t i = 1; i < out->count; i++) {
        ProviderDir key = out->items[i];
        size_t j = i;
        while (j > 0 && wcscmp(out->items[j - 1].provider, key.provider) > 0) {
            out->items[j] = out->items[j - 1];
            j--;
        }
        out->items[j] = key;
    }
}

void data_dirs_resolve(ProviderDirs *out) {
    wchar_t home[MAX_PATH];
    default_home(home, MAX_PATH);
    data_dirs_resolve_with(os_env, home, out);
}

/* Existing dir -> itself; existing file -> its parent (if a directory). */
static wchar_t *watchable_dir(const wchar_t *path) {
    bool is_dir;
    if (!path_exists(path, &is_dir)) {
        return NULL;
    }
    if (is_dir) {
        return wstr_dup(path);
    }
    const wchar_t *sep = wcsrchr(path, L'\\');
    if (!sep || sep == path) {
        return NULL;
    }
    size_t len = (size_t)(sep - path);
    wchar_t *parent = malloc((len + 1) * sizeof(wchar_t));
    if (!parent) {
        return NULL;
    }
    wmemcpy(parent, path, len);
    parent[len] = L'\0';
    if (!path_exists(parent, &is_dir) || !is_dir) {
        free(parent);
        return NULL;
    }
    return parent;
}

static int compare_wide(const void *a, const void *b) {
    return wcscmp(*(wchar_t *const *)a, *(wchar_t *const *)b);
}

void data_dirs_watch_paths_with(EnvLookupFn env, const wchar_t *home, WatchPaths *out) {
    size_t provider_count = ARRAYSIZE(PROVIDERS);
    size_t cap = provider_count * MAX_CANDIDATES;
    out->items = calloc(cap, sizeof(wchar_t *));
    out->count = 0;
    if (!out->items) {
        return;
    }
    for (size_t i = 0; i < provider_count; i++) {
        Candidates c;
        candidates_for(&PROVIDERS[i], env, home, &c);
        for (size_t j = 0; j < c.count; j++) {
            wchar_t *dir = watchable_dir(c.paths[j]);
            if (dir && out->count < cap) {
                out->items[out->count++] = dir;
            } else {
                free(dir);
            }
        }
        candidates_free(&c);
    }
    if (out->count > 1) {
        qsort(out->items, out->count, sizeof(wchar_t *), compare_wide);
        size_t unique = 1;
        for (size_t i = 1; i < out->count; i++) {
            if (wcscmp(out->items[i], out->items[unique - 1]) == 0) {
                free(out->items[i]);
            } else {
                out->items[unique++] = out->items[i];
            }
        }
        out->count = unique;
    }
}

void data_dirs_watch_paths(WatchPaths *out) {
    wchar_t home[MAX_PATH];
    default_home(home, MAX_PATH);
    data_dirs_watch_paths_with(os_env, home, out);
}

void provider_dirs_free(ProviderDirs *dirs) {
    for (size_t i = 0; i < dirs->count; i++) {
        free(dirs->items[i].dir);
    }
    free(dirs->items);
    dirs->items = NULL;
    dirs->count = 0;
}

void watch_paths_free(WatchPaths *paths) {
    for (size_t i = 0; i < paths->count; i++) {
        free(paths->items[i]);
    }
    free(paths->items);
    paths->items = NULL;
    paths->count = 0;
}
